addons/armaos/functions/fnc_terminal_addEventHandler.sqf

```diff
@@ -0,0 +1,76 @@
+// https://community.bistudio.com/wiki/DIK_KeyCodes
+#include "\a3\ui_f\hpp\definedikcodes.inc"
+
+params ["_terminalCtrl"];
+
+/* ---------------------------------------- */
+
+private _result = _terminalCtrl ctrlAddEventHandler
+[
+	"KeyDown", 
+	{
+		params ["_displayorcontrol", "_key", "_shift", "_ctrl", "_alt"];
+
+		//hint format ["OUTPUT CONTROL KEY: %1 \n SHIFT: %2 \n CTRL: %3 \n ALT: %4", _key call BIS_fnc_keyCode, _shift, _ctrl, _alt];
+
+		private _computer = _displayorcontrol getVariable "AE3_computer";
+		private _terminal = _computer getVariable "AE3_terminal";
+		private _terminalApplication = _terminal get "AE3_terminalApplication";
+
+		hint format ["TERMINAL: %1", _terminal];
+
+		private _localGameLanguage = language;
+
+		// we can determine the language of arma 3 but not the language of the keyboard layout
+		// if the language is german, it's obvious, that the keyboard layout is also german (this is not the case, if game language is english)
+		// perhaps we need to provide a CBA setting for changing keyboard layout or allow to change the layout directly in the terminal window
+		private _allowedKeys = createHashMap;
+		if (_localGameLanguage == "German") then
+		{
+			_allowedKeys = [] call AE3_armaos_fnc_terminal_getAllowedKeysGerman;
```

Comment:
Generating the hole hashmap every keypress isn't efficient

Reply:
Do you have any suggestions on how to mitigate or prevent this effect?

Reply:
Loading the hashMap into a variable when the user opens the terminal would be an option

Reply:
I followed your advice and changed the mechanism, so the hashmap ist only created/populated at start and if the keyboard layout changes. On keypress the event handler accesses the hashmap by a reference variable. Commit: https://github.com/y0014984/Advanced-Equipment/pull/90/commits/02b88e6b8a89adfd3382f46213bda7c332acb521